dask_ml/model_selection/_incremental.py

```diff
@@ -459,43 +463,54 @@ def _get_params(self):
 
-    def _get_history_results(self, results):
-        # type: (Results) -> Dict
-        """Construct the CV results.
-
-        Has the following keys:
-
-        * params
-        * test_score
-        * mean_test_score
-        * rank_test_score
-        * mean_partial_fit_time
-        * std_partial_fit_time
-        * mean_score_time
-        * std_score_time
-        * partial_fit_calls
-        * model_id
-        """
-        info, model, history = results
-        key = operator.itemgetter("model_id")
-        hist2 = sorted(history, key=key)
-        return hist2
-
-    def _get_best(self, results, history_results):
-        # type: (Dict, Dict) -> Estimator
-        """Select the best estimator from the set of estimators."""
-        best_model_id = first(results.info)
-        key = operator.itemgetter("model_id")
-        best_index = -1
-        # history_results is sorted by (model_id, partial_fit_calls)
-        # best is the model_id with the highest partial fit calls
-        for k, v in itertools.groupby(history_results, key=key):
-            v = list(v)
-            best_index += len(v)
-            if k == best_model_id:
-                break
-
-        return results.models[best_model_id], best_index
+    def _get_cv_results(self, history, model_hist):
+        cv_results = {}
+        best_scores = {}
+        best_scores = {k: hist[-1]["score"] for k, hist in model_hist.items()}
+
+        cv_results = {}
+        for k, hist in model_hist.items():
+            pf_times = list(toolz.pluck("partial_fit_time", hist))
+            score_times = list(toolz.pluck("score_time", hist))
+            cv_results[k] = {
+                "mean_partial_fit_time": np.mean(pf_times),
+                "mean_score_time": np.mean(score_times),
+                "std_partial_fit_time": np.std(pf_times),
+                "std_score_time": np.std(score_times),
+                "test_score": best_scores[k],
+                "model_id": k,
+                "params": hist[0]["params"],
+                "partial_fit_calls": hist[-1]["partial_fit_calls"],
+            }
+        cv_results = list(cv_results.values())  # list of dicts
+        cv_results = {k: [res[k] for res in cv_results] for k in cv_results[0]}
+
+        # Every model will have the same params because this class uses either
+        # ParameterSampler or ParameterGrid
+        cv_results.update(
+            {
+                "param_" + k: v
+                for params in cv_results["params"]
+                for k, v in params.items()
+            }
+        )
+        cv_results = {k: np.array(v) for k, v in cv_results.items()}
+
+        order = (-1 * cv_results["test_score"]).argsort()
+        cv_results["rank_test_score"] = order.argsort() + 1
+        return cv_results
+
+    def _get_best(self, results, cv_results):
+        scores = {
+            k: v[-1]["score"] for k, v in results.info.items() if k in results.models
```

Comment:
One more time, since I'm still shaky on this: when can we have a model in `results.info` that's not in `results.models`? Under what conditions can we end up in a situation where the highest-scoring model isn't returned to us here?

Reply:
> when can we have a model in results.info that's not in results.models

All the time, since this PR also expands `info` so it includes history on every model at ([source](https://github.com/dask/dask-ml/pull/370/files#diff-0a54ded74c8013caf0588e9a5c879e99R260)). I've added that to https://github.com/dask/dask-ml/pull/370#issue-216493783.

> Under what conditions can we end up in a situation where the highest-scoring model isn't ~~returned to us here~~ in `results.models`?

There are two dimensions to the adaptive search:

1. how many `partial_fit` calls happened before a model was scored
2. how high the score is

For adaptive searches (which includes stop on plateau), there's no guarantee that the highest scoring model be scored with the maximum number of `partial_fit` calls.

For passive searches (e.g., `RandomizedSearchCV`) there's only one dimension: how high the final score is.

Reply:
> there's no guarantee that the highest scoring model be scored with the maximum number of partial_fit calls.

Just checking though, we're still going to return the highest scoring model back to the user at the end though, regardless of how long it was trained, right?

Reply:
Is there anything stopping us from storing the best performing model of each hyper-parameter set that we've scored so far and then returning the best one to the user?

Reply:
Sorry, I should have said for each hyperparameter set being considered.

I would expect that the patience process would keep track of the best scoring model we've seen so far, and if we've not seen anything better after a certain number of partial fit calls, would stop training that model, keeping a reference to the best seen so far.

Over time we would cull our models based on the scores of their best seen score.

Eventually this set would pare down to one.  That would would be the best model that we've seen so far.

In the diagram that you've shown above, I think that would store the model at the observed peaks.  This is how we would manage patience.

Reply:
I don't see an issue with the current implementation, so I'd rather not file an issue. I'll certainly respond if an issue is filed.

> I would expect that the patience process would keep track of the best scoring model we've seen so far, and if we've not seen anything better after a certain number of partial fit calls, would stop training that model, keeping a reference to the best seen so far.

That's not how I'd expect the patience process/stopping training on plateau to work – I'd expect [early stopping], which is what Scikit-Learn's SGDClassifier does ([docs]). I see early stopping as a proxy for convergence, and a method to

* prevent very bad performance
* limit computation

I do not see it as a method to provide marginal performance gains. At best, holding on to the best model will improve the score by `tol`, at least for real-world use.

I don't think this will be an issue in practice: the variance of validation scores on plateau is typically small (e.g., [CIFAR10 errors] from the [torch blog]). I'm inclined to say this noise is (typically) due to randomness in the dataset or training algorithm.

For real-world models/datasets/problems, adaptive algorithms will succeed in returning the highest scoring model, at least in terms of the final fit score (because they retain the highest scoring model every time they're called). The image I showed earlier isn't realistic, and a low probability result where even a passive search wouldn't return a high performing model.

[docs]:http://scikit-learn.org/stable/modules/sgd.html#stopping-criterion
[torch blog]:http://torch.ch/blog/2016/02/04/resnets.html
[CIFAR10 errors]:https://raw.githubusercontent.com/torch/torch.github.io/master/blog/_posts/images/resnets_CIFARtesterror_modeldepth.png
[3]:http://papers.nips.cc/paper/6768-qsgd-communication-efficient-sgd-via-gradient-quantization-and-encoding
[early stopping]:https://en.wikipedia.org/wiki/Early_stopping
[2]:http://scikit-learn.org/stable/modules/sgd.html#stopping-criterion
[1]:https://github.com/scikit-learn/scikit-learn/blob/c13ba26c4780d6088c40db3cc0f8367df4e27f1f/sklearn/linear_model/sag_fast.pyx#L501-L508